### PR TITLE
Histogram: 6 x (1 to 20) x performance improvement

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Based on #624 (**I have set this as base branch for review purposes, do not merge before that branch**). During testing the groupby algorithm the insufficient performance of `histogram` became apparent.

This PR adds the following improvements:
1. Avoid using the iterators of `ViewIndex`. These iterators are slow (in particular for random access). `std::upper_bound` is apparently using those heavily which severely hampered performance. We are making a copy here to avoid this. In principle we could even avoid the copy by just checking for a continuous range, but as long as we support only 1D bin edges this is not relevant. This change brought a 6x performance improvement.
2. Avoid `std::upper_bound` for "linear" bin edges. This replaces it by simple floating point arithmetic, thus avoiding many branches and inefficient memory access. This brings up to 20x performance improvement (for many events per histogram). For few events per histogram I did not notice an improvement. I believe this is actually just because we then end up measuring the allocation/initialization cost of the large 2D output array, so the actual "histogramming" cost is not visible.

Eventually we will likely want to make a similar optimized implementation for logarithmic binning.

### CODE REVIEW

File: docs/getting-started/quick-start.ipynb [hunk 1]
```diff
@@ -20,7 +20,8 @@
     "import scipp as sc\n",
     "from scipp import Dim\n",
     "from scipp.plot import plot\n",
-    "sc.plot.config.backend = 'static'"
+    "sc.plot.config.backend = 'matplotlib:quiet'\n",
```
Comment (OwenArnold) [inline docs/getting-started/quick-start.ipynb]: There are some changes here and below that don't seem relevant to this PR. Accidentally staged?

File: core/dataset_operations.cpp [hunk 2]
```diff
@@ -31,23 +32,40 @@ DataArray histogram(const DataConstProxy &sparse,
...
-            auto it = std::upper_bound(edgesSpan.begin(), edgesSpan.end(), c);
-            if (it != edgesSpan.end() && it != edgesSpan.begin())
-              ++(*(curRes + (--it - edgesSpan.begin())));
+        if (scipp::numeric::is_linspace(edges)) {
+          // Special implementation for linear bins. Gives a 0x to 20x speedup
+          // for few and many events per histogram, respectively.
+          const double offset = edges.front();
+          const double nbin = static_cast<double>(len);
+          const double scale = nbin / (edges.back() - edges.front());
+          for (scipp::index i = 0; i < sparse_.dims().volume(); ++i) {
```
Comment (OwenArnold) [inline core/dataset_operations.cpp]: Looks good.

File: common/test/is_linspace_test.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+TEST(IsLinspaceTest, generate_addition) {
+  std::vector<double> range;
+  double current = 345.4564675;
+  std::generate_n(std::back_inserter(range), 1e5, [&current]() {
+    const double step = 0.0034674;
+    current += step;
+    return current;
+  });
+
+  ASSERT_TRUE(is_linspace(range));
```
Comment (OwenArnold) [inline common/test/is_linspace_test.cpp]: 👍 for decent testing.

File: common/include/scipp/common/numeric.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+template <class Range> bool is_linspace(const Range &range) {
+  if (scipp::size(range) < 2)
+    return false;
+  if (range.back() <= range.front())
+    return false;
+
+  using T = typename Range::value_type;
+  const T delta = (range.back() - range.front()) / (scipp::size(range) - 1);
+  constexpr int32_t ulp = 4;
```
Comment (OwenArnold) [inline common/include/scipp/common/numeric.h]: Would be useful to have the hard-code 4 for Unit of Least Precision explained.

Comment (SimonHeybrock) [reply common/include/scipp/common/numeric.h]: Actually I do not have an explanation, I just put something random and small, assuming that "a handful of operations that change by 1 ulp should be allowed". Do you have a guideline how to choose?

File: docs/getting-started/quick-start.ipynb (hunk 1 above)
Comment (SimonHeybrock) [reply docs/getting-started/quick-start.ipynb]: Thanks. Gone now that I fixed the base branch.

File: common/include/scipp/common/numeric.h (hunk 4 above)
Comment (OwenArnold) [reply common/include/scipp/common/numeric.h]: No I don't have a good suggestion for this choice. Maybe @igudich has a suggestion or method for choosing.
